allennlp/modules/conditional_random_field.py

```diff
@@ -0,0 +1,170 @@
+            emit_scores = inputs[:, i].contiguous().view(batch_size, num_tags, 1)
+            # Transition scores are (next_tag, prev_tag) so we broadcast along the instance axis.
+            transition_scores = self.transitions.view(1, num_tags, num_tags)
+            # Alpha is for the prev_tag, so we broadcast along the next_tag axis.
+            alpha = alpha.view(batch_size, 1, num_tags)
+
+            # We add the emit+transition scores only if timestep i-1 was included.
+            inner = alpha + (emit_scores + transition_scores) * masks[i-1].view(batch_size, 1, 1)
+
+            # Now we log_sum_exp over the "prev_tag" axis.
+            alpha = log_sum_exp(inner)
+
+        # Every sequence needs to end with a transition to the stop_tag.
+        stops = alpha + self.transitions[self.stop_tag].view(1, num_tags)
+
+        # Finally we log_sum_exp along the num_tags dim, result is (batch_size,)
+        return log_sum_exp(stops)
+
+    def _joint_likelihood(self,
+                          inputs: torch.Tensor,
+                          tags: torch.Tensor,
+                          mask: torch.ByteTensor) -> torch.Tensor:
+        """
+        Computes the numerator term for the log-likelihood, which is just score(inputs, tags)
+        """
+        batch_size, sequence_length, num_tags = inputs.data.shape
+        mask = mask.float()
+        masks = [mask[:, i] for i in range(sequence_length)]
+        # contiguous tags
+        ctags = [tags[:, i].contiguous() for i in range(sequence_length)]
+
+        # Start with the transition scores from start_tag to the first tag in each input
+        score = self.transitions.index_select(0, tags[:, 0])[:, self.start_tag]
+
+        # Broadcast the transition scores to one per batch element
+        broadcast_transitions = self.transitions.view(1, num_tags, num_tags).expand(batch_size, num_tags, num_tags)
+
+        # Add up the scores for the observed transitions and all the inputs but the last
+        for i in range(sequence_length - 1):
+            # Each is shape (batch_size,)
+            prev_tag, next_tag = ctags[i], ctags[i+1]
+
+            # The scores for transitioning from prev_tag to next_tag
+            transition_score = (
+                    broadcast_transitions
+                    # Choose the next_tag-th row for each input
+                    .gather(1, next_tag.view(-1, 1, 1).expand(batch_size, 1, num_tags))
+                    # Squeeze down to (batch_size, num_tags)
+                    .squeeze(1)
+                    # Then choose the prev_tag-th column for each of those
+                    .gather(1, prev_tag.view(-1, 1))
+                    # And squeeze down to (batch_size,)
+                    .squeeze(1)
+            )
+
+            # The score for using prev_tag
+            input_score = inputs[:, i].contiguous().gather(1, prev_tag.view(-1, 1)).squeeze(1)
```

Comment:
This is an `emit_score`, isn't it?  Also, I think the `inputs` should be `logits` here, and in the other method.